src/mlpack/methods/mean_shift/mean_shift_impl.hpp

```diff
@@ -260,12 +261,32 @@ inline void MeanShift<UseKernel, KernelType, MatType>::Cluster(
     }
   }
 
-  // Assign centroids to each point.
-  neighbor::KNN neighborSearcher(centroids);
-  arma::mat neighborDistances;
-  arma::Mat<size_t> resultingNeighbors;
-  neighborSearcher.Search(data, 1, resultingNeighbors, neighborDistances);
-  assignments = resultingNeighbors;
+  // If no centroid has converged due to too little iterations and without
+  // forcing convergence, take 1 random centroid calculated.
+  if (centroids.empty())
```

Comment:
We can also optimize for the case of `centroids.n_cols == 1` with another `else if` and call `assignments.zeros()` there to avoid the `KNN` search.

Reply:
Noted, will add this in.